### PR TITLE
Create a class to hold field capabilities for one index.

### PR DESCRIPTION
**Pull request by jtibshirani**
Currently, the same class `FieldCapabilities` is used both to represent the
capabilities for one index, and also the merged capabilities across indices. To
help clarify the logic, this PR proposes to create a separate class
`IndexFieldCapabilities` for the capabilities in one index. The refactor will
also help when adding `source_path` information in #49264, since the merged
source path field will have a different structure from the field for a single index.

Individual changes: 
* Add a new class IndexFieldCapabilities.
* Remove extra constructor from FieldCapabilities.
* Combine the add and merge methods in FieldCapabilities.Builder.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/fieldcaps/IndexFieldCapabilities.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        // Previously we also stored placeholders for the 'indices' arrays.
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+            // Do nothing.
+        } else {
+            in.readOptionalStringArray();
+            in.readOptionalStringArray();
+            in.readOptionalStringArray();
+        }
+
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/IndexFieldCapabilities.java]: The serialization turned out to be a bit messy unfortunately.

**Review by jimczi — APPROVED** [review]
The change makes sense to me, the bwc path is a bit difficult to read (we serialize `IndexFieldCapabilities` as `FieldCapabilities` for nodes < 8) but I don't have a better idea to simplify.

**Comment by jtibshirani** [comment]
Thanks @jimczi for the review.

> (we serialize `IndexFieldCapabilities` as `FieldCapabilities` for nodes < 8)

I pushed a change to the serialization logic to make this clearer, I'm curious if you find it better.

**Review by jimczi — APPROVED** [review]
It looks better to me, thanks @jtibshirani

**Pull request merged**